sdk/storage/Azure.Storage.Common/src/Shared/LazyLoadingReadOnlyStream.cs

```diff
@@ -0,0 +1,221 @@
+            Func<HttpRange, object, bool, CancellationToken, Task<Response<T>>> downloadToAsyncFunc,
+            Func<HttpRange, object, bool, CancellationToken, Response<T>> downloadToFunc,
+            long position = 0,
+            int bufferSize = Constants.DefaultDownloadCopyBufferSize,
+            object requestConditions = default)
+        {
+            _downloadToAsyncFunc = downloadToAsyncFunc;
+            _downloadToFunc = downloadToFunc;
+            _position = position;
+            _bufferSize = bufferSize;
+            _requestConditions = requestConditions;
+        }
+
+        public override int Read(byte[] buffer, int offset, int count)
+            => ReadInternal(
+                async: false,
+                buffer,
+                offset,
+                count,
+                default)
+            .EnsureCompleted();
+
+        public override async Task<int> ReadAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+            => await ReadInternal(
+                async: true,
+                buffer,
+                offset,
+                count,
+                cancellationToken)
+                .ConfigureAwait(false);
+
+        public async Task<int> ReadInternal(bool async, byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+        {
+            ValidateReadParameters(buffer, offset, count);
+
+            if (_stream == null)
+            {
+                await Download(async, cancellationToken).ConfigureAwait(false);
+                if (_lastDownloadBytes == 0)
+                {
+                    return 0;
+                }
+            }
+
+            int totalCopiedBytes = 0;
+            do
+            {
+                int copiedBytes = async
+                    ? await _stream.ReadAsync(buffer, offset, count).ConfigureAwait(false)
+                    : _stream.Read(buffer, offset, count);
+                offset += copiedBytes;
+                count -= copiedBytes;
+                _position += copiedBytes;
+                totalCopiedBytes += copiedBytes;
+
+                // We've run out of bytes in the current block.
+                if (copiedBytes == 0)
+                {
+                    // We hit the end of the blob with the last download call.
+                    if (_position == _length)
+                    {
+                        return totalCopiedBytes;
+                    }
+
+                    // Download the next block
+                    else
+                    {
+                        await Download(async, cancellationToken).ConfigureAwait(false);
+                    }
+                }
+            }
+            while (count > 0);
+            return totalCopiedBytes;
+        }
+
+        private async Task Download(bool async, CancellationToken cancellationToken)
+        {
+            Response<T> response;
+
+            HttpRange range = new HttpRange(_position, _bufferSize);
```

Comment:
Note that we always download _bufferSize, which was specified by the user, or defaulted to 1 MB.  If the user is attempting to read more than _bufferSize, we will have to do multiple downloads for a single `Read()` call.

We could consider changing this behavior, but keep in mind the user could specify up to max int bytes at a time.

Reply:
I have implemented the suggestion to return whatever data we have right away, instead of waiting to download `count` bytes worth of data.

Reply:
Also changed default buffer size to 4 MB.